rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+      return RCL_RET_ACTION_SERVER_TAKE_FAILED; \
+    } \
+    return RCL_RET_ERROR;  /* error already set */ \
+  } \
+  return RCL_RET_OK; \
+
+#define SEND_SERVICE_RESPONSE(Type) \
+  if (!rcl_action_server_is_valid(action_server)) { \
+    return RCL_RET_ACTION_SERVER_INVALID;  /* error already set */ \
+  } \
+  RCL_CHECK_ARGUMENT_FOR_NULL(ros_ ## Type ## _response, RCL_RET_INVALID_ARGUMENT); \
+  rmw_request_id_t request_header;  /* ignored */ \
+  rcl_ret_t ret = rcl_send_response( \
+    &action_server->impl->Type ## _service, &request_header, ros_ ## Type ## _response); \
+  if (RCL_RET_OK != ret) { \
+    return RCL_RET_ERROR;  /* error already set */ \
+  } \
+  return RCL_RET_OK; \
+
+rcl_ret_t
+rcl_action_take_goal_request(
+  const rcl_action_server_t * action_server,
+  void * ros_goal_request)
+{
+  TAKE_SERVICE_REQUEST(goal);
+}
+
+rcl_ret_t
+rcl_action_send_goal_response(
+  const rcl_action_server_t * action_server,
+  void * ros_goal_response)
+{
+  SEND_SERVICE_RESPONSE(goal);
+}
+
+// Implementation only
+static int64_t
+_goal_info_stamp_to_nanosec(const rcl_action_goal_info_t * goal_info)
+{
+  assert(goal_info);
+  return (int64_t)(goal_info->stamp.sec * 1e9) + (int64_t)goal_info->stamp.nanosec;
+}
+
+// Implementation only
+static void
+_nanosec_to_goal_info_stamp(const int64_t * nanosec, rcl_action_goal_info_t * goal_info)
+{
+  assert(nanosec);
+  assert(goal_info);
+  goal_info->stamp.sec = (int32_t) (*nanosec / (int64_t)1e9);
+  goal_info->stamp.nanosec = (uint32_t)(*nanosec - (int64_t)(goal_info->stamp.sec * 1e9));
+}
+
+rcl_action_goal_handle_t *
+rcl_action_accept_new_goal(
+  rcl_action_server_t * action_server,
+  const rcl_action_goal_info_t * goal_info)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return NULL;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_info, NULL);
+
+  // Check if goal with same ID already exists
+  if (rcl_action_server_goal_exists(action_server, goal_info)) {
+    RCL_SET_ERROR_MSG("goal ID already exists");
+    return NULL;
+  }
+
+  // Allocate space in the goal handle pointer array
+  rcl_allocator_t allocator = action_server->impl->options.allocator;
+  rcl_action_goal_handle_t ** goal_handles = action_server->impl->goal_handles;
+  const size_t num_goal_handles = action_server->impl->num_goal_handles;
+  // TODO(jacobperron): Don't allocate for every accepted goal handle,
+  //                    instead double the memory allocated if needed.
+  const size_t new_num_goal_handles = num_goal_handles + 1;
+  void * tmp_ptr = allocator.reallocate(
+    goal_handles, new_num_goal_handles * sizeof(rcl_action_goal_handle_t *), allocator.state);
+  if (!tmp_ptr) {
+    RCL_SET_ERROR_MSG("memory allocation failed for goal handle pointer");
+    return NULL;
+  }
+  goal_handles = (rcl_action_goal_handle_t **)tmp_ptr;
+
+  // Allocate space for a new goal handle
+  tmp_ptr = allocator.allocate(sizeof(rcl_action_goal_handle_t), allocator.state);
```

Comment:
True, the main reason I went with an array of pointers was that is made it easier for removing expired goals from the middle of the list:

https://github.com/ros2/rcl/blob/06df1fdf3cd1789fbf6ae186784bbfde18e7d9f6/rcl_action/src/rcl_action/action_server.c#L533-L535

But I guess it wouldn't bee too much overhead to copy the entire goal handle in this case (instead of just the pointer).